src/components/SetupTab/ArgInput/index.jsx

```diff
@@ -8,159 +8,223 @@ import Row from 'react-bootstrap/Row';
 import InputGroup from 'react-bootstrap/InputGroup';
 import Modal from 'react-bootstrap/Modal';
 
+function FormLabel(props) {
+  return (
+    <Form.Label column sm="3" htmlFor={props.argkey}>
+      {props.children}
+    </Form.Label>
+  );
+}
+FormLabel.propTypes = {
+  argkey: PropTypes.string.isRequired,
+  children: PropTypes.element.isRequired,
+};
 
-export class ArgInput extends React.PureComponent {
-  constructor(props) {
-    super(props)
-  }
+function Feedback(props) {
+  return (
+    <Form.Control.Feedback type="invalid" id={`${props.argkey}-feedback`}>
+      {`${props.argtype} : ${(props.message)}`}
+    </Form.Control.Feedback>
+  );
+}
+Feedback.propTypes = {
+  argkey: PropTypes.string.isRequired,
+  argtype: PropTypes.string.isRequired,
+  message: PropTypes.string,
+};
+Feedback.defaultProps = {
+  message: '',
+};
 
+export default class ArgInput extends React.PureComponent {
   render() {
-    const argkey = this.props.argkey
-    const argSpec = this.props.argSpec
-    let ArgInput;
+    const {
+      argkey,
+      argSpec,
+      handleBoolChange,
+      handleChange,
+      isValid,
+      selectFile,
+      touched,
+      ui_option,
+      validationMessage,
+      value,
+    } = this.props;
+    let Input;
 
     // These types need a text input, and some also need a file browse button
     if (['csv', 'vector', 'raster', 'directory', 'freestyle_string', 'number'].includes(argSpec.type)) {
-      const typeLabel = argSpec.type !== 'freestyle_string' ? argSpec.type : 'string'
-      const labelText = <span>{argSpec.name} <em> ({typeLabel})</em></span>
-      ArgInput = 
-        <Form.Group as={Row} key={argkey} className={'arg-' + this.props.ui_option} data-testid={'group-' + argkey}>
-          <Form.Label column sm="3"  htmlFor={argkey}>
-            {labelText}
-          </Form.Label>
+      const typeLabel = argSpec.type !== 'freestyle_string' ? argSpec.type : 'string';
+      Input = (
+        <Form.Group
+          as={Row}
+          key={argkey}
+          className={`arg-${ui_option}`}
+          data-testid={`group-${argkey}`}
```

Comment:
> Interesting! Is this parameter something that's only used in the tests? Or something else?

@phargogh That's right, it's only used in tests to select this DOM element and assert some content. It's meant as a last resort for selecting nodes, but there are a couple instances where the preferred selectors aren't viable. https://testing-library.com/docs/dom-testing-library/api-queries#bytestid

Here's an excerpt from `setuptab.test.js`

```
// The 'hide' style is applied to the whole Form.Group which 
// includes the Label and Input. Right now, the only good way
// to query the Form.Group node is using a data-testid property.
expect(utils.getByTestId('group-arg3')).toHaveClass('arg-hide')
```

Reply:
Ooh! this brings back memories of UI testing with Cypress. It's weird to me that you have to add a property that's only used for tests, but I remember sometimes it is the only way. Tests that didn't access nodes by their test-id were often flaky, but hopefully that's less of a problem here. I like that you're assigning the `data-testid` programmatically rather than giving it an arbitrary name, that often was messy.